sdk/core/azure_core/src/http/pager.rs

```diff
@@ -2,140 +2,412 @@
-                response,
-                continuation,
-            },
-            None => PagerResult::Complete { response },
+            Some(next) => PagerResult::More { response, next },
+            None => PagerResult::Done { response },
         }
     }
 }
 
-/// Represents a paginated stream of results generated through HTTP requests to a service.
+impl<P, N: fmt::Debug> fmt::Debug for PagerResult<P, N> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::More { next, .. } => f
+                .debug_struct("More")
+                .field("next", &next)
+                .finish_non_exhaustive(),
+            Self::Done { .. } => f.debug_struct("Done").finish_non_exhaustive(),
+        }
+    }
+}
+
+/// Represents a single page of items returned by a collection request to a service.
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait Page {
+    /// The type of items in the collection.
+    type Item;
+    /// The type containing items in the collection e.g., [`Vec<Self::Item>`](Vec).
+    type IntoIter: Iterator<Item = Self::Item>;
+
+    /// Gets a single page of items returned by a collection request to a service.
+    async fn into_items(self) -> crate::Result<Self::IntoIter>;
+}
+
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+impl<P: DeserializeOwned + Page + Send> Page for Response<P> {
+    type Item = P::Item;
+    type IntoIter = P::IntoIter;
+    async fn into_items(self) -> crate::Result<Self::IntoIter> {
+        // TODO: Need to add Format type parameter and adjust accordingly when available.
+        let page: P = self.into_json_body().await?;
+        page.into_items().await
+    }
+}
+
+/// Represents a paginated stream of items returned by a collection request to a service.
 ///
-/// Specifically, this is a [`PageStream`] that yields [`Response<T>`] values.
-pub type Pager<T> = PageStream<Response<T>>;
+/// Specifically, this is a [`ItemIterator`] that yields [`Response<T>`] items.
+pub type Pager<P> = ItemIterator<Response<P>>;
 
-/// Represents a paginated stream of results from a service.
-#[pin_project::pin_project]
-pub struct PageStream<T> {
-    #[pin]
-    #[cfg(not(target_arch = "wasm32"))]
-    stream: Pin<Box<dyn Stream<Item = Result<T, Error>> + Send>>,
+#[cfg(not(target_arch = "wasm32"))]
+type BoxedStream<P> = Box<dyn Stream<Item = Result<P, Error>> + Send>;
 
+#[cfg(target_arch = "wasm32")]
+type BoxedStream<P> = Box<dyn Stream<Item = Result<P, Error>>>;
+
+/// Iterates over a collection of items or individual pages of items from a service.
+///
+/// You can asynchronously iterate over items returned by a collection request to a service,
+/// or asynchronously fetch pages of items if preferred.
+#[pin_project::pin_project]
+pub struct ItemIterator<P: Page> {
```

Comment:
I wonder if this should be `PageItemIterator` or something that specifies that it's an iterator _over items that are retrieved in pages_ as opposed to just an iterator over items? I don't feel super strongly about that though, `ItemIterator` is OK with me if that feels too verbose. The fact that `ItemIterator` takes a `Page` as a generic parameter also helps clue the user in that this is for pagination.

Reply:
Fair. For now, though, I think I prefer the conciseness. If we get feedback indicating confusion, though, I'm certainly not married to the name.